src/main/java/picard/illumina/parser/TileMetricsUtil.java

```diff
@@ -117,18 +122,68 @@ private static File fileTileMetricsFile(File illuminaRunDirectory, Integer numCy
 
             //only create for cluster records
             if (record.isClusterRecord()) {
-                // Snag the phasing data for each read in the read structure. For both types of phasing values, this is the median of all of the individual values seen
-                final Collection<TilePhasingValue> tilePhasingValues = getTilePhasingValues(record.getLaneTileCode(),
-                        phasingMetricsFiles, readStructure, validationStringency);
-
-                tiles.add(new Tile(record.getLaneNumber(), record.getTileNumber(), tileMetricsIterator.getDensity(), record.getMetricValue(),
+                final Collection<TilePhasingValue> tilePhasingValues = phasingValues.get(record.getLaneNumber()).get(record.getTileNumber());
+                tiles.add(new Tile(record.getLaneNumber(), record.getTileNumber(), density, record.getMetricValue(),
                         tilePhasingValues.toArray(new TilePhasingValue[tilePhasingValues.size()])));
             }
         }
-
         return Collections.unmodifiableCollection(tiles);
     }
 
+    private static boolean v3TileMetricsHasClusterRecordsFirstPass(Map<String, ? extends Collection<IlluminaTileMetrics>> locationToMetricsMap) {
+        final Map.Entry<String, ? extends Collection<IlluminaTileMetrics>> entry = locationToMetricsMap.entrySet().iterator().next();
+        final IlluminaTileMetrics record = CollectionUtil.getSoleElement(entry.getValue());
+        return record.isClusterRecord();
+    }
+
+    public static Collection<Tile> parseClusterRecordsFromTileMetricsV3(
+            final Collection<File> tileMetricsOutFiles,
+            final Map<Integer, File> phasingMetricsFiles,
+            final ReadStructure readStructure
+    ) throws FileNotFoundException {
+        Map<Integer, Map<Integer, Collection<TilePhasingValue>>> phasingValues = getTilePhasingValues(phasingMetricsFiles, readStructure);
+        for (File tileMetricsOutFile : tileMetricsOutFiles) {
+            TileMetricsOutReader tileMetricsIterator = new TileMetricsOutReader(tileMetricsOutFile, TileMetricsOutReader.TileMetricsVersion.THREE);
+            final Collection<IlluminaTileMetrics> tileMetrics = determineLastValueForLaneTileMetricsCode(tileMetricsIterator);
+            final Map<String, ? extends Collection<IlluminaTileMetrics>> locationToMetricsMap = partitionTileMetricsByLocation(tileMetrics);
+            if (v3TileMetricsHasClusterRecordsFirstPass(locationToMetricsMap)) {
+                final float density = tileMetricsIterator.getDensity();
+                return getTileClusterRecordsV3(locationToMetricsMap, phasingValues, density);
+            }
+        }
+        for (File tileMetricsOutFile : tileMetricsOutFiles) {
```

Comment:
Could the logic in this loop be combined with the loop above? There's some duplication.